### ISSUE TITLE
Register class not found

### ISSUE THREAD

**Issue by NeatMonster** [quarkslab/QBDI#46]
I encountered the following issue when running QBDI on the `gpg` binary. From what I gather, it's might be some kind of size confusion between the size of all the operands of the instruction.

```
[ExecBlockManager::getProgrammedExecBlock] Cache miss for sequence 0x562d7b4e918e
[Engine::run] Cache miss for 0x562d7b4e918e, patching & instrumenting new basic block
[Engine::patch] Patching basic block at address 0x562d7b4e918e
[Engine::patch] Patching 0x562d7b4e918e         xor        edx, edx
[Engine::patch] Patch rule 14 applied
[Engine::patch] Patch of size 1 generated
[Engine::patch] Patching 0x562d7b4e9190         mov        byte ptr [rax + rdx], cl
[Engine::patch] Patch rule 14 applied
[Engine::patch] Patch of size 1 generated
[Engine::patch] Patching 0x562d7b4e9193         movzx        ecx, byte ptr [rbx + rdx + 1]
[Engine::patch] Patch rule 14 applied
[Engine::patch] Patch of size 1 generated
[Engine::patch] Patching 0x562d7b4e9198         mov        byte ptr [rax + rdx + 1], cl
[Engine::patch] Patch rule 14 applied
[Engine::patch] Patch of size 1 generated
[Engine::patch] Patching 0x562d7b4e919c         add        rdx, 2
[Engine::patch] Patch rule 14 applied
[Engine::patch] Patch of size 1 generated
[Engine::patch] Patching 0x562d7b4e91a0         movzx        ecx, byte ptr [rbx + rdx]
[Engine::patch] Patch rule 14 applied
[Engine::patch] Patch of size 1 generated
[Engine::patch] Patching 0x562d7b4e91a4         test        cl, cl
[Engine::patch] Patch rule 14 applied
[Engine::patch] Patch of size 1 generated
[Engine::patch] Patching 0x562d7b4e91a6         jne        -24
[Engine::patch] Patch rule 9 applied
[Engine::patch] Patch of size 6 generated
[Engine::patch] Basic block starting at address 0x562d7b4e918e ended at address 0x562d7b4e91a6
[Engine::instrument] Instrumenting basic block [0x562d7b4e918e, 0x562d7b4e91a6]
[Engine::instrument] Instrumenting 0x562d7b4e918e         xor        edx, edx
[Engine::instrument] Instrumentation rule 0 applied
[Engine::instrument] Instrumenting 0x562d7b4e9190         mov        byte ptr [rax + rdx], cl
[Engine::instrument] Instrumentation rule 0 applied
[Engine::instrument] Instrumentation rule 2 applied
[Engine::instrument] Instrumentation rule 3 applied
[Engine::instrument] Instrumenting 0x562d7b4e9193         movzx        ecx, byte ptr [rbx + rdx + 1]
[TempManager::getRegSize] Register class not found
[1]    30260 abort (core dumped)  LD_LIBRARY_PATH=/usr/local/lib LD_PRELOAD=./tracer.so gpg --output file.gpg
```

**Comment by haxelion** [comment]
I wrote a test case using exactly this instruction and these operands with memory logging enabled and it works fine. It must be a very tricky corner case related to the exact choice of instrumentations you use.

**Comment by NeatMonster** [comment]
Here is a minimal tracer script that triggers the issue:
```
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#include "QBDIPreload.h"

QBDIPRELOAD_INIT;

static VMAction onInstruction(VMInstanceRef vm, GPRState *gprState,
                              FPRState *fprState, void *data) {
    return QBDI_CONTINUE;
}

static VMAction onMemory(VMInstanceRef vm, GPRState *gprState,
                         FPRState *fprState, void *data) {
    return QBDI_CONTINUE;
}

static VMAction onBasicBlock(VMInstanceRef vm, const VMState *vmState,
                             GPRState *gprState, FPRState *fprState,
                             void *data) {
    return QBDI_CONTINUE;
}

static VMAction onTransferReturn(VMInstanceRef vm, const VMState *vmState,
                                 GPRState *gprState, FPRState *fprState,
                                 void *data) {
    return QBDI_CONTINUE;
}

int qbdipreload_on_run(VMInstanceRef vm, rword start, rword stop) {
    qbdi_addCodeCB(vm, QBDI_PREINST, onInstruction, NULL);
    qbdi_addMemAccessCB(vm, QBDI_MEMORY_READ_WRITE, onMemory, NULL);
    qbdi_addVMEventCB(vm, QBDI_BASIC_BLOCK_NEW, onBasicBlock, NULL);
    qbdi_addVMEventCB(vm, QBDI_EXEC_TRANSFER_RETURN, onTransferReturn, NULL);
    qbdi_run(vm, start, stop);
    return QBDIPRELOAD_NO_ERROR;
}

int qbdipreload_on_exit(int status) {
    return QBDIPRELOAD_NO_ERROR;
}

int qbdipreload_on_start(void *main) {
    return QBDIPRELOAD_NOT_HANDLED;
}

int qbdipreload_on_premain(void *gprCtx, void *fpuCtx) {
    return QBDIPRELOAD_NOT_HANDLED;
}

int qbdipreload_on_main(int argc, char** argv) {
    return QBDIPRELOAD_NOT_HANDLED;
}
```


Here are the commands I typed:
```
rm -f tracer.so file.txt file.gpg
clang tracer.c -fPIC -shared -lQBDI -lQBDIPreload -o tracer.so
echo "Hello world" > file.txt
LD_LIBRARY_PATH=/usr/local/lib LD_PRELOAD=./tracer.so gpg --output file.gpg --encrypt --yes --recipient chubain@quarkslab.com file.txt
```

**Comment by haxelion** [comment]
I cannot reproduce that bug even with that tracer code. Since this is really an assert "that should never happen", I highly suspect that a memory corruption happened somewhere.
On a related note, I also have problems instrumenting gpg on my machine, so it might be linked to something QBDI doesn't like in gpg. I will continue investigate as such.

**Comment by NeatMonster** [comment]
Here is the result of adding modifying the debug message:
```
[TempManager::getRegSize] Register class 0 not found
```

**Comment by haxelion** [comment]
That's weird, getSubReg should not return register 0.
Can you try out with this branch and tell me what happens? https://github.com/quarkslab/QBDI/tree/fix_regclass

**Comment by NeatMonster** [comment]
Here is the result of trying the new branch:

```
>>> LD_LIBRARY_PATH=/usr/local/lib LD_PRELOAD=./tracer.so gpg --output file.gpg --encrypt --yes --recipient chubain@quarkslab.com file.txt 2> output.log
[1]    24252 abort (core dumped)  LD_LIBRARY_PATH=/usr/local/lib LD_PRELOAD=./tracer.so gpg --output file.gpg
>>> tail -80 output.log
[Assembly::writeInstruction] Assembly result at 0x7f809e9cb580 is: 48 8d 8b 82 c3 2f 00
[ExecBlock::newShadow] Registering new tagged shadow 6for instID 3 wih tag 65520
[Assembly::writeInstruction] Assembling 	mov	qword ptr [rip + 3682], rcx at 0x7f809e9cb587
[Assembly::writeInstruction] Assembly result at 0x7f809e9cb587 is: 48 89 0d 62 0e 00 00
[Assembly::writeInstruction] Assembling 	movabs	rbx, 140190378711590 at 0x7f809e9cb58e
[Assembly::writeInstruction] Assembly result at 0x7f809e9cb58e is: 48 bb 26 5a b9 9d 80 7f 00 00
[Assembly::writeInstruction] Assembling 	mov	rcx, qword ptr [rbx + 3130242] at 0x7f809e9cb598
[Assembly::writeInstruction] Assembly result at 0x7f809e9cb598 is: 48 8b 8b 82 c3 2f 00
[ExecBlock::newShadow] Registering new tagged shadow 7for instID 3 wih tag 65522
[Assembly::writeInstruction] Assembling 	mov	qword ptr [rip + 3666], rcx at 0x7f809e9cb59f
[Assembly::writeInstruction] Assembly result at 0x7f809e9cb59f is: 48 89 0d 52 0e 00 00
[Assembly::writeInstruction] Assembling 	mov	rbx, qword ptr [rip + 3419] at 0x7f809e9cb5a6
[Assembly::writeInstruction] Assembly result at 0x7f809e9cb5a6 is: 48 8b 1d 5b 0d 00 00
[Assembly::writeInstruction] Assembling 	mov	rcx, qword ptr [rip + 3420] at 0x7f809e9cb5ad
[Assembly::writeInstruction] Assembly result at 0x7f809e9cb5ad is: 48 8b 0d 5c 0d 00 00
[Assembly::writeInstruction] Assembling 	mov	qword ptr [rip + 3413], rcx at 0x7f809e9cb5b4
[Assembly::writeInstruction] Assembly result at 0x7f809e9cb5b4 is: 48 89 0d 55 0d 00 00
[Assembly::writeInstruction] Assembling 	mov	qword ptr [rip + 3398], rbx at 0x7f809e9cb5bb
[Assembly::writeInstruction] Assembly result at 0x7f809e9cb5bb is: 48 89 1d 46 0d 00 00
[Assembly::writeInstruction] Assembling 	movabs	rbx, 140190378711590 at 0x7f809e9cb5c2
[Assembly::writeInstruction] Assembly result at 0x7f809e9cb5c2 is: 48 bb 26 5a b9 9d 80 7f 00 00
[Assembly::writeInstruction] Assembling 	mov	rcx, qword ptr [rbx + 3130242] at 0x7f809e9cb5cc
[Assembly::writeInstruction] Assembly result at 0x7f809e9cb5cc is: 48 8b 8b 82 c3 2f 00
[Assembly::writeInstruction] Assembling 	mov	qword ptr [rip + 3494], rcx at 0x7f809e9cb5d3
[Assembly::writeInstruction] Assembly result at 0x7f809e9cb5d3 is: 48 89 0d a6 0d 00 00
[Assembly::writeInstruction] Assembling 	mov	rbx, qword ptr [rip + 3367] at 0x7f809e9cb5da
[Assembly::writeInstruction] Assembly result at 0x7f809e9cb5da is: 48 8b 1d 27 0d 00 00
[Assembly::writeInstruction] Assembling 	mov	rcx, qword ptr [rip + 3368] at 0x7f809e9cb5e1
[Assembly::writeInstruction] Assembly result at 0x7f809e9cb5e1 is: 48 8b 0d 28 0d 00 00
[Assembly::writeInstruction] Assembling 	mov	qword ptr [rip + 3353], rbx at 0x7f809e9cb5e8
[Assembly::writeInstruction] Assembly result at 0x7f809e9cb5e8 is: 48 89 1d 19 0d 00 00
[Assembly::writeInstruction] Assembling 	movabs	rbx, 140190389433424 at 0x7f809e9cb5ef
[Assembly::writeInstruction] Assembly result at 0x7f809e9cb5ef is: 48 bb 50 f4 5c 9e 80 7f 00 00
[Assembly::writeInstruction] Assembling 	mov	qword ptr [rip + 3496], rbx at 0x7f809e9cb5f9
[Assembly::writeInstruction] Assembly result at 0x7f809e9cb5f9 is: 48 89 1d a8 0d 00 00
[Assembly::writeInstruction] Assembling 	movabs	rbx, 0 at 0x7f809e9cb600
[Assembly::writeInstruction] Assembly result at 0x7f809e9cb600 is: 48 bb 00 00 00 00 00 00 00 00
[Assembly::writeInstruction] Assembling 	mov	qword ptr [rip + 3487], rbx at 0x7f809e9cb60a
[Assembly::writeInstruction] Assembly result at 0x7f809e9cb60a is: 48 89 1d 9f 0d 00 00
[Assembly::writeInstruction] Assembling 	movabs	rbx, 3 at 0x7f809e9cb611
[Assembly::writeInstruction] Assembly result at 0x7f809e9cb611 is: 48 bb 03 00 00 00 00 00 00 00
[Assembly::writeInstruction] Assembling 	mov	qword ptr [rip + 3478], rbx at 0x7f809e9cb61b
[Assembly::writeInstruction] Assembly result at 0x7f809e9cb61b is: 48 89 1d 96 0d 00 00
[Assembly::writeInstruction] Assembling 	movabs	rbx, 140190393611839 at 0x7f809e9cb622
[Assembly::writeInstruction] Assembly result at 0x7f809e9cb622 is: 48 bb 3f b6 9c 9e 80 7f 00 00
[Assembly::writeInstruction] Assembling 	mov	qword ptr [rip + 3437], rbx at 0x7f809e9cb62c
[Assembly::writeInstruction] Assembly result at 0x7f809e9cb62c is: 48 89 1d 6d 0d 00 00
[Assembly::writeInstruction] Assembling 	mov	rbx, qword ptr [rip + 3278] at 0x7f809e9cb633
[Assembly::writeInstruction] Assembly result at 0x7f809e9cb633 is: 48 8b 1d ce 0c 00 00
[Assembly::writeInstruction] Assembling 	jmp	2198 at 0x7f809e9cb63a
[Assembly::writeInstruction] Assembly result at 0x7f809e9cb63a is: e9 92 08 00 00
[Assembly::writeInstruction] Assembling 	jmp	2193 at 0x7f809e9cb63f
[Assembly::writeInstruction] Assembly result at 0x7f809e9cb63f is: e9 8d 08 00 00
[ExecBlockManager::writeBasicBlock] Sequence 0x7f809db95a20-0x7f809db95a26 written in ExecBlock 0x557bad270670 as seqID 3
[ExecBlockManager::getExpansionRatio] 110546 / 2444
[ExecBlockManager::updateRegionStat] Region 42 has 2189 bytes available of which 2080 are reserved for 46 bytes of untranslated code
[ExecBlockManager::getProgrammedExecBlock] Looking up sequence at address 7f809db95a20
[ExecBlockManager::searchRegion] Searching for address 0x7f809db95a20
[ExecBlockManager::searchRegion] Exact match for region 42 [0x7f809db959f0, 0x7f809db95a36]
[ExecBlockManager::getProgrammedExecBlock] Found sequence 0x7f809db95a20 in ExecBlock 0x557bad270670 as seqID 3
[ExecBlockManager::searchRegion] Searching for address 0x7f809db95a20
[ExecBlockManager::searchRegion] Exact match for region 42 [0x7f809db959f0, 0x7f809db95a36]
[ExecBlock::execute] Executing ExecBlock 0x557bad270670 programmed with selector at 0x7f809e9cb500
[ExecBlock::execute] Execution of ExecBlock 0x557bad270670 resumed at 0x7f809e9cb500
[ExecBlock::makeRX] Making ExecBlock 0x557bad270670 RX
[ExecBlock::execute] Callback request by ExecBlock 0x557bad270670 for callback 0x7f809e5cf430
[ExecBlock::execute] Callback 0x7f809e5cf430 returned CONTINUE
[ExecBlock::execute] Execution of ExecBlock 0x557bad270670 resumed at 0x7f809e9cb568
[ExecBlock::makeRX] Making ExecBlock 0x557bad270670 RX
[ExecBlock::execute] Callback request by ExecBlock 0x557bad270670 for callback 0x7f809e5cf450
[ExecBlock::execute] Callback 0x7f809e5cf450 returned CONTINUE
[ExecBlock::execute] Execution of ExecBlock 0x557bad270670 resumed at 0x7f809e9cb63f
[ExecBlock::makeRX] Making ExecBlock 0x557bad270670 RX
[Engine::run] Next address to execute is 0x7f809d171df0
[ExecBroker::getReturnPoint] Found instrumented return address on the stack at 0x7ffec7160ed8
[Engine::run] Executing 0x7f809d171df0 through execBroker
[ExecBroker::getReturnPoint] Found instrumented return address on the stack at 0x7ffec7160ed8
[ExecBroker::transferExecution] Patched 0x7ffec7160ed8 hooking return address 0x7f809dba0718 with 0x7f809e9f5ed1
[ExecBroker::transferExecution] Transfering execution to 0x7f809d171df0 using transferBlock 0x557bad24e1d8
[ExecBlock::makeRX] Making ExecBlock 0x557bad24e1d8 RX
```

**Comment by NeatMonster** [comment]
That actually seems to have fixed the issue. I could encrypt and decrypt a simple text file successfully.

**Comment by nezetic** [comment]
Fix merged from PR #50.
